### ISSUE TITLE
Better type suggest for array of object literals

### ISSUE THREAD

**Issue by dmitry-vsl** [facebook/flow#107]
When I run `flow suggest` on the following code
```
/* @flow */
var users = [ {name: 'Alice'}, {name: 'Bob'} ];
```
it offers the following type for `users` variable:
```
Array<{name: string;} | {name: string;}>
```
I think it can be just
```
Array<{name: string;}>
```

**Comment by samwgoldman** [comment]
So this ultimately comes down to desired semantics. The current behavior of flow's inference for array literals leads me to believe that your example should produce `Array<{ name: string }>`, as you expect.

Here's how it works today, based on my understanding of the source code.
1. Flow infers a type for each element of the array literal
2. Flow turns that list of inferred types into a set, presumably intending to collapse equiv types
3. If the cardinality of that set is 1, we create a simple array type, otherwise we create a tuple

It seems that the equivalence function for the set in step 2 is able to detect that two `number`s are the same, but it doesn't compare object types based on their properties.

FWIW, the equivalence function is called `reasonless_compare` in Constraint_js.

**Comment by samwgoldman** [comment]
Huh, how did that comment unassign @lordhoto?

**Comment by lordhoto** [comment]
Good question, maybe it did that because I am not an organization member anymore?

**Comment by facebook-github-bot-2** [comment]
Thank you for reporting this issue and appreciate your patience. We've  notified the core team for an update on this issue. We're looking for a response within the next 30 days or the issue may be closed.

**Comment by mroch** [comment]
Looks like this was sorta resolved by #925, where we now print this example as `[{name: string}, {name: string}]` instead of `Array<{name: string} | {name: string}>`, since it's a tuple.

Please reopen if you come up with another example.
